### PR TITLE
SONARIAC-1461 Parsing error on `@` character

### PR DESCRIPTION
**Pull request by petertrr**

### CODE REVIEW

File: iac-extensions/spring-config/src/main/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessor.java [hunk 1]
```diff
@@ -53,4 +54,12 @@ private static String removeInlineCommentAfterDocumentBreak(String line) {
     }
     return line;
   }
+
+  /**
+   * Surround Maven variable substitution in quotes instead of '@' to avoid parsing issues.
+   * We don't resolve these variables, so we shouldn't lose information by doing this.
+   */
+  private static String transformMavenSubstitutions(String line) {
+    return MAVEN_VARIABLE_AT_TOKEN_START.matcher(line).replaceAll("$1'${name}'$3");
```
Comment (petertrr) [inline iac-extensions/spring-config/src/main/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessor.java:63]: I was thinking about either substituting `@...@` with `${,,,}`, but that would require tracking text range adjustments, or tracking the information about the substituted symbols and replacing them back in the converter. However, I don't see an immediate value from it, because it's not likely that maven variable name will clash with a sensitive property value, where we would raise an issue.

File: iac-common/src/main/java/org/sonar/iac/common/api/tree/impl/TextPointer.java
Comment (rudy-regazzoni-sonarsource) [inline iac-common/src/main/java/org/sonar/iac/common/api/tree/impl/TextPointer.java:1]: Nice improvement 👍 Records are truly a thing.

File: iac-extensions/spring-config/src/main/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessor.java [hunk 2]
```diff
@@ -27,23 +28,23 @@ public class SpringConfigYamlPreprocessor {
   private static final Pattern LINE_SEPARATORS = Pattern.compile("(\n|\r|\r\n|\u2028|\u2029)");
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/spring-config/src/main/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessor.java:28]: I'm expecting this regex to not work as expected.
This would match `\r` before it has a chance to match `\r\n`, transforming `data1\r\ndata2` into `['data1', '\ndata2']`.
This is not related to this PR, so it's probably better to handle it in a separate ticket/PR.
```suggestion
  private static final Pattern LINE_SEPARATORS = Pattern.compile("(\r\n|\n|\r|\u2028|\u2029)");
```

File: iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessorTest.java [hunk 3]
```diff
@@ -68,4 +68,29 @@ void shouldRemoveInlineCommentsAfterDocumentBreak() {
...
+      bar: prefix-@maven.property@-suffix
+      baz: |
+        @maven.property@
+      """;
+    // language=yaml
+    var expected = """
+      foo: 'maven.property'
+      foo2: 'maven.property'-suffix
+      foo3: '@maven.property@'
+      bar: prefix-@maven.property@-suffix
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessorTest.java:88]: What about a line with two variables?
```
      bar: prefix-@maven.property1@-@maven.property2@
```

File: iac-extensions/spring-config/src/main/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessor.java (hunk 1 above)
Comment (rudy-regazzoni-sonarsource) [reply iac-extensions/spring-config/src/main/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessor.java:63]: Yes, it's probably better not to mess with offset. This is generally not so fun to handle, as shown with our docker preprocessor.
Seems good to me to keep it as it is for now, until we get additional values from implementing a more complete/complex solution.

File: iac-extensions/spring-config/src/main/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessor.java [hunk 4]
```diff
@@ -27,23 +28,23 @@ public class SpringConfigYamlPreprocessor {
...
 
-    if (Stream.of(lines).noneMatch(SpringConfigYamlPreprocessor::containsInlineCommentAfterDocumentBreak)) {
+    if (Stream.of(lines).noneMatch(containsInlineCommentAfterDocumentBreak.or(containsMavenVariableAtTokenStart))) {
       return source;
     }
 
     return Stream.of(lines)
       .map(SpringConfigYamlPreprocessor::removeInlineCommentAfterDocumentBreak)
+      .map(SpringConfigYamlPreprocessor::transformMavenSubstitutions)
       .collect(Collectors.joining("\n"));
```
Comment (rudy-regazzoni-sonarsource) [inline iac-extensions/spring-config/src/main/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessor.java:45]: We lose the original line separator.
This should not be an issue, although I still want to raise awareness about it (this was an issue for docker, where the preprocessor was changing the offset, changing a `\r\n` into `\n` would have higher impact).

**Review by rudy-regazzoni-sonarsource — APPROVED** [review]
LGTM! I added a few comments, maybe the multiple variable thing would require some modification if it's a valid concern. It can then be tackled in a different ticket/PR, I let you take the decision on it.

File: iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessorTest.java (hunk 3 above)
Comment (petertrr) [reply iac-extensions/spring-config/src/test/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessorTest.java:88]: Good catch, I'll add a test case. The idea is to only substitute the first variable in a line, because when `@` is in the middle of a value, snakeyaml doesn't have a problem with it.

File: iac-extensions/spring-config/src/main/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessor.java (hunk 2 above)
Comment (petertrr) [reply iac-extensions/spring-config/src/main/java/org/sonar/iac/springconfig/parser/yaml/SpringConfigYamlPreprocessor.java:28]: Good catch! I added a quick test case for it as well to this PR.
